argon2_remote_support.go

```diff
@@ -0,0 +1,460 @@
+			return &Argon2RemoteOutput{
+				Command:     input.Command,
+				ErrorType:   Argon2RemoteErrorUnexpectedInput,
+				ErrorString: "cannot supply keylen for \"time\" command",
+			}
+		}
+
+		duration, err := InProcessArgon2KDF().Time(input.Mode, costParams)
+		if err != nil {
+			return &Argon2RemoteOutput{
+				Command:     input.Command,
+				ErrorType:   Argon2RemoteErrorUnexpected,
+				ErrorString: fmt.Sprintf("cannot run time command: %v", err),
+			}
+		}
+		return &Argon2RemoteOutput{
+			Command:  input.Command,
+			Duration: duration,
+		}
+	default:
+		return &Argon2RemoteOutput{
+			Command:     input.Command,
+			ErrorType:   Argon2RemoteErrorInvalidCommand,
+			ErrorString: fmt.Sprintf("invalid command: %q", string(input.Command)),
+		}
+	}
+}
+
+// WaitAndRunArgon2RequestInRemoteProcess waits for a [Argon2RemoteInput] request on the
+// supplied io.Reader before running it and sending a [Argon2RemoteOutput] response back via
+// the supplied io.Writer. These will generally be connected to the process's os.Stdin and
+// os.Stdout when using - certainly when using [NewRemoteArgon2KDF] on the parent side.
+// This function can only be called once in a process. Subsequent calls in the same
+// process will result in an error response being returned.
+func WaitAndRunArgon2RequestInRemoteProcess(in io.Reader, out io.Writer) error {
+	var input *Argon2RemoteInput
+	dec := json.NewDecoder(in)
+	dec.DisallowUnknownFields()
+	if err := dec.Decode(&input); err != nil {
+		return fmt.Errorf("cannot decode input: %w", err)
+	}
+
+	output := RunArgon2RequestInRemoteProcess(input)
+
+	enc := json.NewEncoder(out)
+	if err := enc.Encode(output); err != nil {
+		return fmt.Errorf("cannot encode output: %w", err)
+	}
+
+	return nil
+}
+
+// remoteArgon2KDFImpl is an Argon2KDFImpl that runs the KDF in a remote process,
+// using the remote JSON protocol defined in this package.
+type remoteArgon2KDFImpl struct {
+	newRemoteCommand func() (*exec.Cmd, error)
+}
+
+func (k *remoteArgon2KDFImpl) runInRemoteProcess(params *Argon2RemoteInput) (res *Argon2RemoteOutput, err error) {
+	cmd, err := k.newRemoteCommand()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create remote process: %w", err)
+	}
+
+	stdinPipe, err := cmd.StdinPipe()
+	if err != nil {
+		// This doesn't fail once the OS pipe is created, so there's no
+		// cleanup to do on failure paths.
+		return nil, fmt.Errorf("cannot create stdin pipe: %w", err)
+	}
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		// This doesn't fail once the OS pipe is created, so there's no
+		// cleanup to do on failure paths.
+		return nil, fmt.Errorf("cannot create stdout pipe: %w", err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("cannot start remote process: %w", err)
+	}
+	defer func() {
+		// Run Cmd.Wait in a defer so that we shut down on error paths too,
+		// and we capture the Wait error if there was no other error.
+		waitErr := cmd.Wait()
```

Comment:
do we need a timeout mechanism around this?

Reply:
I've refactored this quite a bit now. Not only is there a timeout mechanism when waiting for the remote process to exit (and go 1.20 introduces a nicer way to do this), there is also an optional watchdog mechanism that exercises the command processing / response sending code on the remote side, so that the parent knows that it's eventually going to get a response and can terminate the operation if this fails.